# Analysis Challenge Assignment 2
#### Author: Madeline Maeloa

## Import and Clean

In [2]:
# Import necessary modules
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression

In [3]:
# Import training data
trainingData = pd.read_csv("aca2_dataset_training.csv")
trainingData

UNIQUEID SCHOOL Class  GRADE CODER  STUDENTID  Gender  OBSNUM  \
0         34880      A   T8T      2     Z     965160       1      12   
1         32344      B   T9U      3     Z      11665       1      11   
2         14774      B   T9Q      0     Z     187441       0      21   
3         19468      A   T8S      1     Z     402837       0      17   
4         31570      B   T9T      2     Z     300647       1       2   
...         ...    ...   ...    ...   ...        ...     ...     ...   
22179     30762      B   T9Q      0     Z     749212       1       5   
22180     14611      B   T9Q      0     Z      87701       0      12   
22181     20432      A  T8VQ      4     Z     130859       0       6   
22182     38264      E   T6Q      0     Y     874565       1      10   
22183     21734      A   T8R      4     Z     468563       0       3   

       totalobs-forsession     Activity ONTASK  TRANSITIONS  NumACTIVITIES  \
0                      224  Wholecarpet      Y            1              2   
1                      225   Individual      Y            2              3   
2                      374   Smallgroup      Y            3              4   
3                      320   Wholedesks      Y            5              6   
4                       35   Wholedesks      Y            2              3   
...                    ...          ...    ...          ...            ...   
22179                   95  Wholecarpet      Y            2              3   
22180                  211   Smallgroup      N            3              4   
22181                   95   Smallgroup      N            1              2   
22182                  147   Wholedesks      Y            1              2   
22183                   49   Wholedesks      Y            0              1   

       FORMATchanges  NumFORMATS     Obsv/act  Transitions/Durations  \
0                  1           2  1390.500000               0.000664   
1                  2           3   890.333333               0.002545   
2                  1           2   770.500000               0.001439   
3                  0           1   592.333333               0.001407   
4                  1           2   870.000000               0.005882   
...              ...         ...          ...                    ...   
22179              2           3   969.000000               0.002959   
22180              1           2   770.500000               0.001439   
22181              0           1  1060.000000               0.000472   
22182              1           2  1204.500000               0.000838   
22183              0           1  2076.000000               0.000000   

       Total Time  
0            1027  
1             213  
2            2078  
3            3229  
4             260  
...           ...  
22179         652  
22180         708  
22181         362  
22182          60  
22183         803  

[22184 rows x 18 columns]

In [4]:
# Creates a new table called classData with variables to include in classifier
classData = trainingData[['NumACTIVITIES','ONTASK']]
classData

NumACTIVITIES ONTASK
0                  2      Y
1                  3      Y
2                  4      Y
3                  6      Y
4                  3      Y
...              ...    ...
22179              3      Y
22180              4      N
22181              2      N
22182              2      Y
22183              1      Y

[22184 rows x 2 columns]

In [5]:
# Prepare the data set by creating a dummy variable for ONTASK
#dummy = pd.get_dummies(classData['ONTASK'], prefix = 'ONTASK')
#dummy

In [6]:
# Merging the two tables to drop the ONTASK and ONTASK_N column
#classDataMerged = pd.concat([classData, dummy], axis=1)
#classDataMerged = classDataMerged.drop(['ONTASK', 'ONTASK_N'], axis=1)
#classDataMerged

## Descriptive Data

In [7]:
# Descriptive statistics for NumACTIVITIES
classData.describe()

NumACTIVITIES
count   22184.000000
mean        3.383114
std         1.315163
min         1.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         7.000000

In [9]:
# Counting the number of 'yes' and 'no' in the ONTASK column
classData['ONTASK'].value_counts()

Y    14938
N     7246
Name: ONTASK, dtype: int64

## Naive Bayes Model

In [10]:
# Import GaussianNB from sklearn and createts Xs and Y
from sklearn.naive_bayes import GaussianNB

Xs_NB = classData.drop('ONTASK', axis = 1)
ONTASK = classData['ONTASK']

In [11]:
# Initialize the classification naive Bayes model 
classNBModel = GaussianNB()
classNBModel.fit(Xs_NB, ONTASK)

GaussianNB()

In [12]:
# Use predict() to use given Xs to predict the given Y
# Use list comprehension to see if each element in certified_pred matches elements in certified
# Results return true (1) or false (0)
# Take summation of them and divide by length of the list to get accuracy 
ONTASK_pred = classNBModel.predict(Xs_NB)
performance = [item in ONTASK_pred for item in ONTASK]
print('The accuracy is', sum(performance)/len(performance)*100, '%')

The accuracy is 67.3368193292463 %


## Model Evaluation

In [13]:
# Import validation data
vData = pd.read_csv("aca2_dataset_validation.csv")
vData

UNIQUEID SCHOOL Class  GRADE CODER  STUDENTID  Gender  OBSNUM  \
0        20774      A  T8VQ      4     Z      15210       0       1   
1        30961      B   T9S      1     Z     165165       0       1   
2        35767      A   T8R      4     Z     188548       1       1   
3        19990      A   T8T      2     Z     224646       0       1   
4        18668      A   T8S      1     Z     402837       0       1   
...        ...    ...   ...    ...   ...        ...     ...     ...   
5542     21660      A   T8R      4     Z     857436       0      31   
5543     21662      A   T8R      4     Z     402069       1      31   
5544     21665      A   T8R      4     Z      94671       1      31   
5545     21682      A   T8R      4     Z     762145       1      32   
5546     21683      A   T8R      4     Z     753929       1      32   

      totalobs-forsession     Activity ONTASK  TRANSITIONS  NumACTIVITIES  \
0                       1  Wholecarpet      Y            1              2   
1                       1   Wholedesks      Y            3              4   
2                       1  Wholecarpet      Y            3              4   
3                       1  Wholecarpet      N            2              3   
4                       1   Individual      Y            3              4   
...                   ...          ...    ...          ...            ...   
5542                  487      Testing      Y            0              1   
5543                  489      Testing      Y            0              1   
5544                  492      Testing      N            0              1   
5545                  509      Testing      Y            0              1   
5546                  510      Testing      Y            0              1   

      FORMATchanges  NumFORMATS  Obsv/act  Transitions/Durations  Total Time  
0                 1           2   1677.00               0.000446           0  
1                 3           4    826.00               0.003672           0  
2                 2           3    862.75               0.004658           0  
3                 0           1   1073.00               0.000621           0  
4                 2           3   1106.75               0.002841           0  
...             ...         ...       ...                    ...         ...  
5542              0           1   2530.00               0.000000        2367  
5543              0           1   2530.00               0.000000        2375  
5544              0           1   2530.00               0.000000        2389  
5545              0           1   2530.00               0.000000        2522  
5546              0           1   2530.00               0.000000        2525  

[5547 rows x 18 columns]

In [14]:
# Create new table validataionData with NumACTIVITIES and ONTASK variables
validationData = vData[['NumACTIVITIES','ONTASK']]
validationData

NumACTIVITIES ONTASK
0                 2      Y
1                 4      Y
2                 4      Y
3                 3      N
4                 4      Y
...             ...    ...
5542              1      Y
5543              1      Y
5544              1      N
5545              1      Y
5546              1      Y

[5547 rows x 2 columns]

#### Step 1: split data into 2 sets

In [16]:
from sklearn.model_selection import train_test_split
Xs_NB = validationData.drop('ONTASK', axis = 1)
ONTASK_NB = validationData['ONTASK']
xs_NB_training, xs_NB_test, y_NB_training, y_NB_test = train_test_split(Xs_NB, ONTASK_NB, test_size = 0.2)

#### Step 2: retrain model with just training data

In [17]:
classNBModel = GaussianNB()
classNBModel.fit(xs_NB_training, y_NB_training)

GaussianNB()

#### Step 3: feed the Xs in the testing dataset and obtain the predicted Ys

In [18]:
# Predicted value in naive bayes model
ONTASK_pred_NB = classNBModel.predict(xs_NB_test)

#### Step 4: compare the predicted Ys with what is actually in the testing dataset (the ground truth) and obtain the confusion matrix

In [19]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_NB_test, ONTASK_pred_NB))

[[  0 356]
 [  0 754]]


#### Step 5: obtain accuracy score

In [194]:
from sklearn.metrics import accuracy_score
ac_NB = accuracy_score(y_NB_test, ONTASK_pred_NB)
print("The accuracy for naive bayes", ac_NB*100, "%")

The accuracy for naive bayes 67.38738738738739 %


## Interpretation of Results

#### For this analysis challenge, I used a Naive Bayes model to train a classifier. Because I do not have any prior programming knowledge, I am still learning and processing the information from ICE 4. Thus, for this assignment, I wanted to keep it simple so that I can focus on the programming part of it so that it doesn’t get too overwhelming. I used the number of activities in one session (NumACTIVITIES) to predict on task (ONTASK) behavior. The reason behind this is that I predicted that the more activities there are in a session, the less the children would be on task. After importing the training data set and running some descriptive statistics, I initialized the classification Naive Bayes model. 

#### I imported the validation data set to evaluate the model that I just created. First, I split the data into 2 sets, then retrained the model, used the Xs to obtain predicted Ys, and then compared the predicted Ys with the actual Ys to get a confusion matrix. The accuracy score of my model is 67.39%. This is lower than what I was expecting. This probably means that it is a better predictor of on task behavior versus just randomly guessing (around 50%), but still less accurate than a blind guess (around 70%). What’s good is that the model works and can help us predict the children’s on task behavior. However, I do think that I should have used more than 1 predictor in the classification model to maybe obtain a higher accuracy score next time. For example, I can add the format of activity one session has taken (NumFORMATS) into the classification model. 